images/virtualization-artifact/pkg/controller/service/lock_service.go

```diff
@@ -0,0 +1,166 @@
+	"github.com/deckhouse/virtualization/api/core/v1alpha2/cvicondition"
+	"github.com/deckhouse/virtualization/api/core/v1alpha2/vdcondition"
+	"github.com/deckhouse/virtualization/api/core/v1alpha2/vicondition"
+)
+
+type LockService struct {
+	client client.Client
+}
+
+func NewLockService(
+	client client.Client,
+) *LockService {
+	return &LockService{
+		client: client,
+	}
+}
+
+func (s LockService) CanUnlockVirtualDiskByImage(ctx context.Context, vd *virtv2.VirtualDisk) bool {
+	log, _ := logger.GetDataSourceContext(ctx, "LockService")
+
+	var cvis virtv2.ClusterVirtualImageList
+	err := s.client.List(ctx, &cvis)
+	if err != nil {
+		log.Error(fmt.Sprintf("failed to list cvi: %s", err))
+		return false
+	}
+
+	for _, cvi := range cvis.Items {
+		ready, _ := GetCondition(cvicondition.ReadyType, cvi.Status.Conditions)
+		if ready.Status == metav1.ConditionTrue {
+			continue
+		}
+
+		if cvi.Spec.DataSource.Type != virtv2.DataSourceTypeObjectRef {
+			continue
+		}
+
+		ref := cvi.Spec.DataSource.ObjectRef
+
+		if ref == nil || ref.Kind != virtv2.ClusterVirtualImageObjectRefKindVirtualDisk {
+			continue
+		}
+
+		if ref.Name == vd.GetName() && ref.Namespace == vd.GetNamespace() {
+			return false
+		}
+	}
+
+	var vis virtv2.VirtualImageList
+	err = s.client.List(ctx, &vis, &client.ListOptions{Namespace: vd.Namespace})
+	if err != nil {
+		log.Error(fmt.Sprintf("failed to list vi: %s", err))
+		return false
+	}
+
+	for _, vi := range vis.Items {
+		ready, _ := GetCondition(vicondition.ReadyType, vi.Status.Conditions)
+		if ready.Status == metav1.ConditionTrue {
+			continue
+		}
+
+		if vi.Spec.DataSource.Type != virtv2.DataSourceTypeObjectRef {
+			continue
+		}
+
+		ref := vi.Spec.DataSource.ObjectRef
+
+		if ref == nil || ref.Kind != virtv2.VirtualImageObjectRefKindVirtualDisk {
+			continue
+		}
+
+		if ref.Name == vd.GetName() {
+			return false
+		}
+	}
+
+	return true
+}
+
+func (s LockService) LockVirtualDiskByImage(ctx context.Context, vd *virtv2.VirtualDisk) error {
+	inUseCondition, ok := GetCondition(vdcondition.InUseType, vd.Status.Conditions)
+
+	if !ok || inUseCondition.Status == metav1.ConditionUnknown {
+		inUseCondition = metav1.Condition{
+			Type:   vdcondition.InUseType,
+			Reason: vdcondition.InUseByVirtualImage,
+			Status: metav1.ConditionTrue,
+		}
+
+		SetCondition(inUseCondition, &vd.Status.Conditions)
+		err := s.client.Status().Update(ctx, vd)
+		if err != nil {
+			return err
+		}
+	} else if inUseCondition.Reason != vdcondition.InUseByVirtualImage {
+		return NewVirtualDiskUsedByVirtualMachineError(vd.Name)
+	}
+
+	return nil
+}
+
+func (s LockService) UnlockVirtualDisk(ctx context.Context, vd *virtv2.VirtualDisk) error {
+	inUseCondition, ok := GetCondition(vdcondition.InUseType, vd.Status.Conditions)
+	if ok && inUseCondition.Status == metav1.ConditionTrue {
+		inUseCondition = metav1.Condition{
+			Type:   vdcondition.InUseType,
+			Status: metav1.ConditionUnknown,
```

Comment:
It's important not to fill reason as well, to Unknown value
Condition without reason leads to error during Update